File: Blind 75/07 - Trees/JavaScript/226-invert-binary-tree.js
const { performance } = require('perf_hooks');

class TreeNode {
  constructor(val, left, right) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

class Solution {
  method(root) {
    const dfs = (node) => {
      if (!node) return null;

      [node.right, node.left] = [dfs(node.left), dfs(node.right)];

      return node;
    };

    return dfs(root);
  }

  reference(root) {
    const isBaseCase = root === null;
    if (isBaseCase) return root;

    return this.dfs(root);
  }

  dfs = (root) => {
    const left = this.reference(root.left);
    const right = this.reference(root.right);

    root.left = right;
    root.right = left;

    return root;
  };

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(input));
        else this.method(input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(input));
        else this.reference(input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  new TreeNode(
    4,
    new TreeNode(2, new TreeNode(1), new TreeNode(3)),
    new TreeNode(7, new TreeNode(6), new TreeNode(9))
  ),
  new TreeNode(2, new TreeNode(1), new TreeNode(3)),
  null,
];
test.quantify(testCases);

File: Blind 75/07 - Trees/226-invert-binary-tree.py
'''
Given the root of a binary tree, invert the tree, and return its root.
'''

from time import time
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        def dfs(node):
            if not node:
                return

            node.left, node.right = dfs(node.right), dfs(node.left)

            return node

        return dfs(root)

    def reference(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return None

        # Swap the children
        tmp = root.left
        root.left = root.right
        root.right = tmp

        self.reference(root.left)
        self.reference(root.right)
        return root

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.invertTree(case))
                else:
                    self.invertTree(case)
        print(f'Runtime for our solution: {time() - sol_start}\n')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case))
                else:
                    self.reference(case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [
        TreeNode(
            4,
            TreeNode(2, TreeNode(1), TreeNode(3)),
            TreeNode(7, TreeNode(6), TreeNode(9)),
        ),
        TreeNode(2, TreeNode(1), TreeNode(3)),
        None,
    ]
    test.quantify(test_cases)
